assets/app/view/game/part/track.rb

```diff
@@ -28,14 +28,17 @@ def render
           # Array<Array<Path>>
           @routes_paths = @routes.map { |route| route.paths_for(@tile.paths) }
 
-          sorted = (@tile.paths + @tile.stubs)
-            .map { |path| [path, index_for(path)] }
+          paths_and_stubs = @tile.paths + @tile.stubs
+          path_indexes = paths_and_stubs.map { |p| [p, indexes_for(p)] }.to_h
+
+          sorted = paths_and_stubs
+            .map { |path| path_indexes[path].map { |i| [path, i] } }.flatten(1)
```

Comment:
why do you only need to flatten one level?  why not just do a flat_map?

Reply:
The previous line was this:

```
.map { |path| [path, index_for(path)] }
```

That produced an Array with elements that were a 2-length array containing a path and the route (index) that path was used by, like this:

```
[[pathA, 0], [pathB, nil], [pathC, 0], [pathD, 1]]
```

Now that a path can belong to more than one route, I replaced the `index_for` call (which returned an integer) at that point with a reference to the `path_indexes` hash, giving an array of integers. After my `map` call, the data structure looks like this:

```
[[[pathA, 0], [pathA, 1]], [[pathB, nil]], [[pathC, 0], [pathC, 2]], [[pathD, 1]]]
```

Flattening that once produces a structure like the initial one, where we get a list of path+index pairs:

```
[[pathA, 0], [pathA, 1], [pathB, nil], [pathC, 0], [pathC, 2], [pathD, 1]]
```

Flattening all the way would break up the pairs:

```
[pathA, 0, pathA, 1, pathB, nil, pathC, 0, pathC, 2] pathD, 1]
```

Reply:
ok, can you just do path_and_stubs.flat_map {|path| path.map... }

Reply:
OK, I see, I mistakenly thought flat_map flattened all the way

2b1cf270 fixes that, and another error, I had this happen:

![Screenshot from 2020-10-25 16-53-30](https://user-images.githubusercontent.com/1045173/97121407-a1a0f680-16e3-11eb-8100-20a9809b4e88.png)